doc/man-source/irtestcase.1
.TH irtestcase "1" "Last change: Oct 2015" "irtestcase @version@" "User Commands"
.SH NAME
.P
\fBirtestcas\fRe - Create or check decoding testcases.
.SH SYNOPSIS
.P
\fBirtestcase\fR [\fI-l lircrc-file -p prog\fR] <\fIsocket\fR>
.P
\fBirtestcase\fR [\fI-l lircrc-file -p prog\fR] -t \fItestdata\fR <\fIsocket\fR>
.P
\fBirtestcase\fR [\fI-v\fR|\fI-h\fR]

.SH DESCRIPTION
.P
\fBirtestcase\fR is a tool which can record keypresses from a remote together
with the decoded values. Given output from another session, it can
also verify that the decoding still works as expected.
.P
In the first form irtestcase will wait for input and create log files
for it. Doing so, it also echoes received buttons on stdout. There is
no way to stop the program in a clean way, just use ctrl-C when completed.
.P
In the second form it will feed the data from the testdata file to lircd
and log the the decoded data as usually. Thus, this makes it possible to
test the output from the remote which generated the testdata.  In this
form lircd must be running with the "file" driver.
.P
The optional socket is the lircd output socket (lircd must be running). It
defaults to LIRC_SOCKET_PATH in environment, falling back to a hardcoded
default value (@varrundir@/lirc/lircd).

.SH OPTIONS

.TP 4
\fB-l, --lircrc\fR  \fI<lircrc config file>\fR
Log also applications strings (as ircat(1)) using lircrc type config file.

.TP 4
\fB-p, --prog\fR \fI<program id>\fR
When using -l/--lircrc, the program name matching the PROG = entries
in lircrc

.TP 4
\fB-t, --testdata\fR
Use data i file instead of "live" input from the remote. This can be
the durations.log file from a previous run, see FILES. It can also
be logged output from mode2(1).

.TP 4
\fB-v , --version\fR
Print version and exit.

.TP 4
\fB-h , --help\fR
Print help message.

.SH ENVIRONMENT
irtestcase respects LIRC_SOCKET_PATH which if defined holds the
path to the lircd output socket to listen to.

.SH FILES
.TP 4
.B /tmp/irtestcase
irtestcase creates files in /tmp/irtestcase (hardcoded, sorry). After a
successful run here is:
.RS 4
.TP 4
.B durations.log
Printable space/pulse durations in the same format as mode2(1). This file
is usable as -t/--testdata input.
.TP 4
.B codes.log
The decoded  buttons, in same format as irw(1).
.TP 4
.B app_strings.log
The application strings, as presented by ircat(1).
.RE
.TP 4
.B ~/.cache/irtestcase.log
Debug log. Setting the XDG_CACHE_HOME environment variable relocates this
file to $XDG_CACHE_HOME/irtestcase.log. The amount of logging respects the
LIRC_LOGLEVEL environment variable, defaulting to the \fidebug\fR value
in the \fi[lircd]\fR section of lirc_options.conf.

.SH "SEE ALSO"
.P
The documentation for \fBlirc\fR
is maintained as html pages. They are located under html/ in the
documentation directory.


NEWS
0.9.4-devel
  * The caraca driver has been retired (still available in attic/).
  * New driver girs for the GIRS protocol, gives LIRC Arduino support.
    See new info in driver documentation.
  * Enhanced lircd-setup tool (#123)
    - Integrated mode2 and irw tests of current configuration.
    - Documentation links.
    - More consistent data based on new database.
    - Multiple bugfixes.
  * Several patches to make mode2 work with all drivers.
  * Configuration database overhaul (#123, partly)
    - Database lives as yaml files in configs/.
    - Data is collected from drivers, remotes and static files.
    - Better ways for drivers to enumerate devices.
    - There is an unstable python database module used by
      all tools.
  * devinput.lircd.conf is generated at build time to match host kernel.
  * A new script lircd-setup runs as a one-shot service before lircd,
    performing setup commands as root (#121).

  * CLI options for mode2, xmode2 and irrecord has been streamlined.
    All tools falls back to lircd defaults where it makes sense.
  * New logging API with hooks for enhanced messages and filtering
    (also #129).
  * The out-of-tree build interface is enhanced with hooks for external
    drivers to rebuild documentation and add a config file.
  * irrecord has benn taught to handle remotes without gap (#98).
  * Add an "auto" device to devinput driver for single-device
    usecase (#134).

  * Fix buggy handling of not decoded buttons in irrecord (#146).
  * Fix --plugindir not being respected (#144).
  * Fix crash for non-existing device (#145).
  * Fix gid and supplementary groups in CLI apps dropping root.
  * Fix a crash in irrecord when invoked without options.
  * Don't require write access if not needed in default.c (#133).
  * Clean up docs, remove circular links (#136). Documentation is
    now only in lirc.org/html.

0.9.3  4/09/15
  * Digest of 0.9.3pre1..0.9.3pre3

  * New build dependency: xsltproc (sometimes coming with libxml).
  * The irexec semantics have been changed so that the invoked
    programs run disconnected from the main irexec process.
    See irexec(1) (#92).
  * irrecord has got some love:
     + It handles multiple definitions of same symbol OK (#78)
     + Warns about single-key config files (#77)
     + Checks for ambient light disturbances (#76)
     + New option allows updating of existing template file.
     + Template handling bug fixed
     + -l --list-namespace option parsing bug fixed (#101).
     + Minor fixes to irrecord-ui including -k option parsing fixed.
  * mode2 has been updated
     +  handle also LIRCCODE devices (#94),
     +  Drop root privileges (#91)
     +  Add  new scope-like display
     +  Handle the--driver-options switch
  * xmode2 has been updated to handle udp driver, the new
    --driver-options switch + some general overhaul.
  * irrecord has been split into a library and a cli interface
    (#36). The library interface is unstable.

  * The select() calls in lircd, lircrcd and library routines have
    been replaced with poll() to avoid crashes when file descriptors
    becomes > 1024 (#109).
  * Set supplementary groups using effective-user (RH bz 1253907).
  * Work around kernel bug in /dev/lirc0 output (#119).
  * lircmd: Several option parsing bugs fixed, one of which
    blocked the systemd service (#104, RH bz 1234155).
  * lirc-lsremotes now displays logs after parsing files, making
    more parsing warnings visible.
  * lirc-config-tool is fixed to comply with guidelines (#105).
  * Bug in irexec when using modes fixed (#102).
  * Create log directory if required (#124).
  * lirc-setup: Fixed name clash on recent python versions.

  * irtoy: New support for debug LEDs. New documentation
    file, support in lirc-setup and driver table doc added.
  * The udp driver is patched to allow adjustable clockrates,
    see new udp_driver.html driver doc.
  * The mplay driver has been updated with patches which has
    been available "out there".

  * The library routines for parsing lircd.conf now warns for
    duplicate keys and values (#107).
  * Enhanced pkg-config support for drivers compiled out-of-tree
    (see driver-api in the manual).
  * Hardcoded CFLAGS in Makefiles fixed (#124).
  * The compiler has been switched to c++, and filenames are
    updated to *.cpp (#79)
  * Build: now uses parallel make by default.
  * Overall code formatting is improved with more consistent
    layout and updated git hook. See CONTRIBUTE.md

  * Unlinked and hard to find driver docs and manpages have
    been linked.
  * The documentation is heavily revised. Old obsolete text is
    removed, the manual is reorganized, some info is moved to
    manpages and a lot of minor updates and bugfixes.

  * The lircd.simsend and lircd.simreceive tools have been removed,
    superseeded by irsimsend and irsimreceive.
  * A new kernel driver irpipe has been added. This allows feeding data
    from userspace to lircd, which can use the /dev/irpipeX device
    as a clean substitute for /dev/lircX. A companion CLI tool
    irpipe(1) is also added.
  * Options parsing (notably -U) and just listing first remote
    bugs in irsimsend have been fixed.


0.9.3pre3 16/08/15
  * The select() calls in lircd, lircrcd and library routines have
    been replaced with poll() to avoid crashes when file descriptors
    becomes > 1024 (#109).
  * Set supplementary groups using effective-user (RH bz 1253907).
  * Work around kernel bug in /dev/lirc0 output (#119).
  * Create log directory if required (#124).
  * The remotes/ directory is not used any more, removed.
  * The lircd.simsend and lircd.simreceive tools have been removed,
    superseeded by irsimsend and irsimreceive.
  * Minor fixes to lirc-setup (bbd4ccf, e384375).
  * Minor fixes to irrecord-ui including -k option parsing fixed.
  * Hardcoded CFLAGS in Makefiles fixed (#124).
  * Enhanced pkg-config support for drivers compiled out-of-tree
    (see driver-api in the manual).


tools/irtestcase.cpp
/****************************************************************************
** irtestcase.c ************************************************************
****************************************************************************
*
* irtestcase - Log synced streams of raw durations, codes and app strings.
*
*/

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/un.h>
#include <errno.h>
#include <getopt.h>

#include "lirc_client.h"
#include "lirc_private.h"

static const char* const USAGE =
	"Synopsis:\n"
	"irtestcase [-p prog -l lircrc] [-t testdata] <socket>\n"
	"irtestcase [ħ | -v]\n\n"
	"<socket> is the socket connecting to lircd. Defaults to a hardcoded\n"
	"default value, usually /var/run/lirc/lircd. Respects LIRC_SOCKET_PATH in\n"
	"environment.\n\n"
	"Options:\n"
	"   -l  lircrc  Log also translated symbols using lircrc type config file.\n"
	"   -p  prog    Program name used to match entries in lircrc.\n"
	"   -t  path    Use testdata from path.\n"
	"   -v  version Print version.\n"
	"   -h  help    Print this message.\n";

static struct option opts[] = {
	{ "prog",     required_argument, NULL, 'p' },
	{ "lircrc",   required_argument, NULL, 'l' },
	{ "testdata", required_argument, NULL, 't' },
	{ "help",     no_argument,	 NULL, 'h' },
	{ "version",  no_argument,	 NULL, 'v' },
	{ 0,	      0,		 0,    0   }
};

#define LOGDIR          "/tmp/irtestcase"
#define DEVICE_LOG      LOGDIR  "/durations.log"
#define CODE_LOG        LOGDIR  "/codes.log"
#define APP_LOG         LOGDIR  "/app_strings.log"

#define DEFAULT_PROG    "no_such_prog"

/** Delay (us) before sending testdata so we are listening when it comes.*/
static const int TESTDATA_DELAY = 1000000;

static const char* opt_testdata = NULL;
static const char* opt_lircrc = NULL;
static const char* opt_prog = DEFAULT_PROG;

static FILE* app_log = NULL;
static FILE* code_log = NULL;


/** Configure lircd to log received data from driver in path. */
static void set_devicelog(int fd, const char* path)
{
	int r;
	lirc_cmd_ctx command;

	unlink(path);
	lirc_command_init(&command, "SET_INPUTLOG %s\n", path);
	r = lirc_command_run(&command, fd);
	if (r != 0) {
		fprintf(stderr, "Cannot set lircd device log %s\n", path);
		exit(2);
	}
}


/** Configure lircd (file driver) to read testdata from path. */
static void set_testinput(int fd, const char* path)
{
	int r;
	lirc_cmd_ctx command;

	lirc_command_init(&command, "DRV_OPTION set-infile %s\n", path);
	r = lirc_command_run(&command, fd);
	if (r != 0) {
		fputs("Cannot set test input file\n", stderr);
		exit(2);
	}
}


/** Setup and clear the hardcoded output directory. */
static void init_testdir(void)
{
	unlink(CODE_LOG);
	unlink(APP_LOG);
	unlink(DEVICE_LOG);

	mkdir(LOGDIR, 0755);
	if (access(LOGDIR, F_OK) != 0) {
		fprintf(stderr, "Cannot create log directory: %s\n", LOGDIR);
		exit(2);
	}
	code_log = fopen(CODE_LOG, "w");
	if (code_log == NULL) {
		fprintf(stderr, "Cannot open %s, giving up\n", CODE_LOG);
		exit(2);
	}
	if (opt_lircrc != NULL) {
		app_log = fopen(APP_LOG, "w");
		if (app_log == NULL) {
			fprintf(stderr,
				"Cannot open %s, giving up\n", APP_LOG);
			exit(2);
		}
	}
}


/** Get next code from lircd. */
static int nextcode(int fd, char* buff, ssize_t size)
{
	int i;

	i = read(fd, buff, size);
	if (i == -1) {
		perror("read");
		exit(errno);
	}
	;
	if (strstr(buff, "__EOF") != NULL) {
		puts("Exit on EOF");
		exit(0);
	}
	if (i >= 0)
		buff[i] = '\0';
	return i > 0 ? 1 : 0;
}


/**  Send testdata after delay. This is fork(), so nothing comes back. */
static void send_later(int fd, const char* path)
{
	int r;

	r = fork();
	if (r == 0) {
		usleep(TESTDATA_DELAY);
		set_testinput(fd, path);
		exit(0);
	} else if (r < 0) {
		perror("Cannot fork(!)");
		exit(3);
	} else {
		//parent, just proceed.
	}
}


/** Run testcase. */
static int irtestcase(int fd_io, int fd_cmd)
{
	int r;
	struct lirc_config* config;
	char code[64];
	char* c;

	if (opt_lircrc != NULL) {
		if (lirc_readconfig_only(opt_lircrc, &config, NULL) != 0) {
			fputs("Cannot initiate lircrc decoding\n", stderr);
			exit(2);
		}
	}
	while (nextcode(fd_io, code, sizeof(code)) == 1) {
		fputs(code, stdout);
		if (strstr(code, "__EOF") != NULL)
			exit(0);
		fputs(code, code_log);
		if (opt_lircrc != NULL) {
			r = lirc_code2char(config, code, &c);
			while (r == 0 && c != NULL) {
				printf("    %s\n", c);
				fprintf(app_log, "%s\n", c);
				r = lirc_code2char(config, code, &c);
			}
			fflush(app_log);
		}
		fflush(stdout);
		fflush(code_log);
	}
	return 0;
}


int main(int argc, char* argv[])
{
	int fd_io;
	int fd_cmd;
	const char* socketpath;
	char path[128];
	int c;
	const loglevel_t level = options_get_app_loglevel("irtestcase");

	while ((c = getopt_long(argc, argv, "hl:p:t:v", opts, NULL)) != EOF) {
		switch (c) {
		case 'l':
			opt_lircrc = optarg;
			break;
		case 'p':
			opt_prog = optarg;
			break;
		case 't':
			opt_testdata = optarg;
			break;
		case 'h':
			puts(USAGE);
			return EXIT_SUCCESS;
		case 'v':
			printf("%s\n", "irtestcase " VERSION);
			return EXIT_SUCCESS;
		case '?':
			fprintf(stderr, "unrecognized option: -%c\n", optopt);
			fputs("Try `irtestcase --help'.\n", stderr);
			return EXIT_FAILURE;
		}
	}
	if (argc > optind + 1) {
		fputs("irtestcase: Too many arguments (max one).\n", stderr);
		fputs("Try `irtestcase --help'.\n", stderr);
		return EXIT_FAILURE;
	}
	if (strcmp(opt_prog, DEFAULT_PROG) != 0 && opt_lircrc == NULL) {
		fputs("--prog requires --lircrc/-l. Giving up.\n", stderr);
		return EXIT_FAILURE;
	}
	if (opt_lircrc != NULL && strcmp(opt_prog, DEFAULT_PROG) == 0) {
		fputs("--lircrc requires --prog/-p. Giving up.\n", stderr);
		return EXIT_FAILURE;
	}

	init_testdir();
	socketpath = argc == optind + 1 ? argv[optind] : NULL;
	socketpath = socketpath ? socketpath : getenv("LIRC_SOCKET_PATH");
	socketpath = socketpath ? socketpath : LIRCD;
	fd_cmd = lirc_get_local_socket(socketpath, 1);
	if (fd_cmd < 0) {
		fputs("Cannot open lircd socket.\n", stderr);
		exit(3);
	}
	set_devicelog(fd_cmd, DEVICE_LOG);
	if (opt_testdata != NULL)
		send_later(fd_cmd, opt_testdata);

	lirc_log_get_clientlog("irtestcase", path, sizeof(path));
	lirc_log_set_file(path);
	lirc_log_open("irtestcase", 1, level);

	setenv("LIRC_SOCKET_PATH", socketpath, 1);
	fd_io = lirc_init(opt_prog, 1);
	if (fd_io < 0) {
		fputs("Cannot run lirc_init.\n", stderr);
		exit(3);
	}

	return irtestcase(fd_io, fd_cmd);
}
